### PR TITLE
WIP: refactor CI completely

### PR DESCRIPTION
**Pull request by lucacasonato**
I am going to try to split CI into distinct "build" and "test" steps.
Additionally I'll attempt to shard the tests across multiple concurrent
testing VMs.

TODOs:

- [ ] build in sysroot on linux
- [ ] wpt upload
- [ ] run benchmarks
- [ ] upload benchmarks
- [ ] upload release (notes and binaries)

### CODE REVIEW

**Comment by AaronO** [comment]
Long overdue and lots of improvements to make, looking forward to it !

**Comment by bartlomieju** [comment]
Integration tests are split so that each file produces a separate binary - this could be done upfront in a separate PR

**Comment by bartlomieju** [comment]
Current blocker: packaging artifacts on Windows hangs.

**Review by bartlomieju — COMMENTED** [review]
Note to self: changes to files outside of `.github` and `tools/ci` directories should be landed in separate PRs to make it easier to review changes in this PR. In other words: changes to `js_unit_test`, FFI tests and WPT tests could be landed in separate PRs

File: cli/tests/jsunit_tests.rs [hunk 1]
```diff
@@ -0,0 +1,38 @@
+mod integration;
```
Comment (bartlomieju) [inline cli/tests/jsunit_tests.rs:1]: This is effectively a duplicate of `cli/tests/js_unit_tests.rs`, it should be removed

File: tools/wpt/utils.ts [hunk 2]
```diff
@@ -14,12 +14,26 @@ export const {
   ["auto-config"]: autoConfig,
   ["inspect-brk"]: inspectBrk,
   binary,
+  shard: shardStr,
 } = parse(Deno.args, {
   "--": true,
   boolean: ["quiet", "release", "no-interactive", "inspect-brk"],
-  string: ["json", "wptreport", "binary"],
+  string: ["json", "wptreport", "binary", "shard"],
```
Comment (bartlomieju) [inline tools/wpt/utils.ts:21]: I can't find any reference to this arg, either I made a mistake during rebases, this arg is supported internally by WPT suites or it's not working as expected

File: tools/ci/_util.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
+export const ARCHIVE_COUNT = 4;
```
Comment (bartlomieju) [inline tools/ci/_util.js:1]: Having this const seems brittle, figure out were is it used and can we make it smarter

File: tools/ci/_util.js [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+  constructor(path) {
+    const manifestText = Deno.readTextFileSync(path);
+    const manifest = manifestText.split("\n").filter((line) => line.length > 0)
+      .map((line) => JSON.parse(line));
+    this.#artifacts = manifest
+      .filter((item) => item.reason === "compiler-artifact");
+  }
+
+  get bins() {
```
Comment (bartlomieju) [inline tools/ci/_util.js:14]: Add docstrings to all the getters

**Comment by dsherret** [comment]
We've refactored the existing CI and it's faster so I'm going to close this one. Let's land incremental improvements to the CI on top of that work.

**Pull request closed without merging**